fairlearn/reductions/grid_search_classification.py

```diff
@@ -0,0 +1,98 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import copy
+import numpy as np
+
+from fairlearn.metrics.demographic_parity import DemographicParity
+
+
+class GridSearchClassification:
+    def __init__(self,
+                 learner,
+                 fairness_metric=DemographicParity(),
+                 number_of_lagrange_multipliers=7):
+        self.learner = learner
+        if not isinstance(fairness_metric, DemographicParity):
+            raise RuntimeError("DemographicParity is only currently supported fairness metric")
+        self.fairness_metric = fairness_metric
+        self.number_of_lagrange_multipliers = number_of_lagrange_multipliers
+
+    def fit(self, X, Y, protected_attribute, lagrange_multipliers=None):
+        # Verify we have a binary classification problem
+        unique_labels = np.unique(Y)
+        if not set(unique_labels).issubset({0, 1}):
+            raise RuntimeError("Supplied Y labels are not 0 or 1")
+
+        # Extract required statistics from protected_attribute
+        p0, p1, a0_val = self._generate_protected_attribute_info(
+            protected_attribute)
+
+        # If not supplied, generate array of trial lagrange multipliers
+        if lagrange_multipliers is None:
+            limit = 1
+            if p1 > 0 and p0 / p1 > 1:
+                limit = p0 / p1
+            lagrange_multipliers = np.linspace(-2 * limit,
+                                               2 * limit,
+                                               self.number_of_lagrange_multipliers)
+
+        self.all_models = []
+        for current_multiplier in lagrange_multipliers:
+            # Generate weights array
+            sample_weights = self._generate_weights(
+                Y, protected_attribute, current_multiplier, p1 / p0, a0_val)
+
+            # Generate Y'
+            def f(x): return 1 if x > 0 else 0
+            re_labels = np.vectorize(f)(sample_weights)
+
+            # Run the learner
+            current_learner = copy.deepcopy(self.learner)
+            current_learner.fit(
+                X, re_labels, sample_weight=np.absolute(sample_weights))
+
+            # Append the new model, along with its current_multiplier value
+            # to the result
+            # Note that we call it a model because it is a learner which has
+            # had 'fit' called
+            self.all_models.append({"model": current_learner,
+                                    "lagrange_multiplier": current_multiplier})
+
+        # Designate a 'best' model
+        # Selection algorithm not yet fully implemented
+        self.best_model = self.all_models[0]
+
+    def predict(self, X):
```

Comment:
Implement `predict`, `predict_proba`, and their `posterior_` variants as agreed. Should the `posterior_` ones also have the information about each model attached?